dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/user/hibernate/HibernateUserStore.java

```diff
@@ -355,4 +357,27 @@ public int getUserCount()
         Query<Long> query = getTypedQuery( "select count(*) from User" );
         return query.uniqueResult().intValue();
     }
+
+    @Override
+    public User getUser( long id )
+    {
+        return sessionFactory.getCurrentSession().get( User.class, id );
+    }
+
+    @Override
+    public UserCredentials getUserCredentialsByUsername( String username )
+    {
+        if ( username == null )
+        {
+            return null;
+        }
+
+        String hql = "from UserCredentials uc where uc.username = :username";
+
+        javax.persistence.Query query = sessionFactory.getCurrentSession().createQuery( hql );
+        query.setParameter( "username", username );
+        query.setHint( QueryHints.CACHEABLE, true );
+
+        return ( UserCredentials ) query.getResultList().stream().findFirst().orElse( null );

```

Comment:
I think that calling `stream` on ResultList is not super-efficient.
This kind of Utility function normally does the trick:
```
public  <T> T getSingleResult(TypedQuery<T> query) {
        query.setMaxResults(1);
        List<T> list = query.getResultList();
        if (list == null || list.isEmpty()) {
            return null;
        }

        return list.get(0);
    }
```

In order to use it, you need to change line 377 to:
```
javax.persistence.TypedQuery<UserCredentials> query = sessionFactory.getCurrentSession().createQuery( hql, UserCredentials.class );
```

Reply:
Implemented